arbnode/transaction_streamer.go

```diff
@@ -337,80 +401,108 @@ func (s *TransactionStreamer) AddMessagesAndEndBatch(pos arbutil.MessageIndex, f
 	return s.addMessagesAndEndBatchImpl(pos, force, messages, batch)
 }
 
-func (s *TransactionStreamer) addMessagesAndEndBatchImpl(pos arbutil.MessageIndex, force bool, messages []arbstate.MessageWithMetadata, batch ethdb.Batch) error {
+func (s *TransactionStreamer) getPrevPrevDelayedRead(pos arbutil.MessageIndex) (uint64, error) {
 	var prevDelayedRead uint64
 	if pos > 0 {
 		prevMsg, err := s.GetMessage(pos - 1)
 		if err != nil {
-			return fmt.Errorf("failed to get previous message: %w", err)
+			return 0, fmt.Errorf("failed to get previous message for pos %d: %w", pos, err)
 		}
 		prevDelayedRead = prevMsg.DelayedMessagesRead
 	}
 
-	dontReorgAfter := len(messages)
-	afterCount := pos + arbutil.MessageIndex(len(messages))
-	broadcasterQueuedMessagesPos := arbutil.MessageIndex(atomic.LoadUint64(&s.broadcasterQueuedMessagesPos))
-	if afterCount >= broadcasterQueuedMessagesPos {
-		if int(afterCount-broadcasterQueuedMessagesPos) < len(messages) {
-			messages = append(messages, s.broadcasterQueuedMessages[afterCount-broadcasterQueuedMessagesPos:]...)
-		}
-		s.broadcasterQueuedMessages = s.broadcasterQueuedMessages[:0]
-		atomic.StoreUint64(&s.broadcasterQueuedMessagesPos, 0)
-	}
+	return prevDelayedRead, nil
+}
 
+func (s *TransactionStreamer) skipDuplicateMessages(prevDelayedRead uint64, pos arbutil.MessageIndex, messages []arbstate.MessageWithMetadata, quiet bool) (bool, uint64, arbutil.MessageIndex, []arbstate.MessageWithMetadata, error) {
 	reorg := false
-	// Skip any messages already in the database
 	for {
 		if len(messages) == 0 {
 			break
 		}
 		key := dbKey(messagePrefix, uint64(pos))
 		hasMessage, err := s.db.Has(key)
 		if err != nil {
-			return err
+			return false, 0, 0, nil, err
 		}
 		if !hasMessage {
 			break
 		}
 		haveMessage, err := s.db.Get(key)
 		if err != nil {
-			return err
+			return false, 0, 0, nil, err
 		}
 		nextMessage := messages[0]
 		wantMessage, err := rlp.EncodeToBytes(nextMessage)
 		if err != nil {
-			return err
+			return false, 0, 0, nil, err
 		}
 		if bytes.Equal(haveMessage, wantMessage) {
 			// This message is a duplicate, skip it
 			prevDelayedRead = nextMessage.DelayedMessagesRead
 			messages = messages[1:]
 			pos++
-			dontReorgAfter--
 		} else {
-			var dbMessageParsed arbstate.MessageWithMetadata
-			err := rlp.DecodeBytes(haveMessage, &dbMessageParsed)
-			if err != nil {
-				log.Warn("TransactionStreamer: Reorg detected! (failed parsing db message)", "pos", pos, "err", err)
-			} else {
-				var gotHeader *arbos.L1IncomingMessageHeader
-				if nextMessage.Message != nil {
-					gotHeader = nextMessage.Message.Header
+			reorg = true
+			if !quiet {
+				var dbMessageParsed arbstate.MessageWithMetadata
+				err := rlp.DecodeBytes(haveMessage, &dbMessageParsed)
+				if err != nil {
+					log.Warn("TransactionStreamer: Reorg detected! (failed parsing db message)", "pos", pos, "err", err)
```

Comment:
Why not log if it's a feed reorg? That still seems bad

Reply:
I guess probably because it's spammy as per `nextAllowedReorgLog`. I'd be in favor of adding a similar timer for this.

Reply:
:+1: